File: python/performances/membership_lookup.py
"""멤버십 조회: list vs set

- 동작: 특정 값이 컬렉션 안에 있는지 확인
- 차이: list는 O(N), set은 평균 O(1)
"""

from __future__ import annotations

import random
import time


def contains_with_list(data: list[int], queries: list[int]) -> int:
    count = 0
    for q in queries:
        if q in data:  # O(N)
            count += 1
    return count


def contains_with_set(data: list[int], queries: list[int]) -> int:
    data_set = set(data)  # O(N)
    count = 0
    for q in queries:
        if q in data_set:  # 평균 O(1)
            count += 1
    return count


def main() -> None:
    n = 400_000
    q = 200_000
    data = list(range(n))
    queries = [random.randint(0, n * 2) for _ in range(q)]

    s = time.perf_counter()
    c1 = contains_with_list(data, queries)
    t1 = time.perf_counter() - s

    s = time.perf_counter()
    c2 = contains_with_set(data, queries)
    t2 = time.perf_counter() - s

    print(f"list 결과={c1}, 시간={t1:.4f}s")
    print(f"set  결과={c2}, 시간={t2:.4f}s")


if __name__ == "__main__":
    main()

File: javascript/performances/membership_lookup.js
function randomInt(max) { return Math.floor(Math.random() * (max + 1)); }

function containsWithList(data, queries) {
  let count = 0;
  for (const q of queries) if (data.includes(q)) count += 1;
  return count;
}

function containsWithSet(data, queries) {
  const dataSet = new Set(data);
  let count = 0;
  for (const q of queries) if (dataSet.has(q)) count += 1;
  return count;
}

const n = 400000;
const q = 200000;
const data = Array.from({ length: n }, (_, i) => i);
const queries = Array.from({ length: q }, () => randomInt(n * 2));

let s = process.hrtime.bigint();
const c1 = containsWithList(data, queries);
let t1 = Number(process.hrtime.bigint() - s) / 1e9;

s = process.hrtime.bigint();
const c2 = containsWithSet(data, queries);
let t2 = Number(process.hrtime.bigint() - s) / 1e9;

console.log(`list 결과=${c1}, 시간=${t1.toFixed(4)}s`);
console.log(`set  결과=${c2}, 시간=${t2.toFixed(4)}s`);
